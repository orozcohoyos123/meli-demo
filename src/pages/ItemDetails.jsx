import React, { useEffect, useState } from "react";
import Categories from "../components/Categories";
import ProductDetails from "../components/ProductDetails";
import productsAdapter from '../adapters/products';
import { useParams } from 'react-router-dom';
import { initialState } from '../initialState';
import "../styles/Common.scss";

/**
 * Component for showing the product details, render the products and categories components
 * 
 * @component
 * @example
 * return (<ItemDetails />)
 */
const ItemDetails = () => {
  /**
  * Hook useParams to get the product id from the url 
  */
  let { itemId } = useParams();

  /**
   * Hook useState loading, set a flag to indicate a loading state
   * @example setQuery(true)
   */
  const [loading, setLoading] = useState(false);

  /**
  * Hook useState product, set the product information to render in a child component
  * @example setProduct({...productDetails})
  */
  const [product, setProduct] = useState(initialState.products[0]);
  
  /**
  * Hook useState category, set the categories array information to render in a child component
  * @example setCategories([{...category}, {...category}])
  */
  const [category, setCategories] = useState(initialState.categories);

   /**
   * Hook useEffect to call the function getProductById from the adapter,
   * it's activated when product id const change 
   */
  useEffect(() => {
    setLoading(true);

    productsAdapter.getProductById(itemId)
      .then(({ item, categories }) => {
        item !== undefined ? setProduct(item) : setProduct(initialState.products[0]);
        categories.length > 0 ? setCategories(categories) : setCategories(initialState.categories);
        setLoading(false);
      })
      .catch(err => {
        //console.error(err);
        setLoading(false);
      });
  }, [itemId]);


  return (
    <main className="view-wrapper">
      <div className="view-detail row">
        <Categories items={category} />
        <ProductDetails product={product} loading={loading} />
      </div>
    </main>
  );
}

export default ItemDetails;