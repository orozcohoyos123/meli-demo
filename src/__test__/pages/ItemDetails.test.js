import * as React from 'react';
import { mount } from 'enzyme';
import { create } from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import productsAdapter  from '../../adapters/products';
import ItemDetails from '../../pages/ItemDetails';
import { act } from "@testing-library/react";

const mockProducts = {
    author: {
        name: "Sebastian",
        lastname: "Orozco"
    },
    item: {
        id: "SRSA_A12312S_123ASD",
        title: "Disco SSD Samsung EVO 3200",
        price: {
            currency: "$",
            amount: 19200,
            decimals: 2,
        },
        picture: "https://http2.mlstatic.com/D_NQ_NP_2X_874785-MCO44525297153_012021-F.webp",
        condition: "Nuevo",
        free_shipping: true,
        sold_quantity: 210,
        description: "Sellado de f√°brica.",
    }
};

describe('<ItemDetails />', () => {
    test('Debe renderizar el componente.', () => {

        const useEffectspy = jest.spyOn(React, "useEffect").mockResolvedValue(mockProducts);

        // bug en jest spyON, no sobreescribe la funcion por lo que no puede espiarla => https://github.com/facebook/jest/issues/6972
        //const spy = jest.spyOn(productsAdapter, 'getProductById'); 
        
        const itemDetails = mount(
            <MemoryRouter>
                <ItemDetails />
            </MemoryRouter>
        );
        
        expect(useEffectspy).toHaveBeenCalled();
        //expect(spy).toHaveBeenCalled();
        expect(itemDetails.length).toEqual(1);
    })

/* 
    test("Debe renderizar el componente simulando useEffect con un mock de productos", async () => {
        await act(async () => {
            jest.spyOn(productsAdapter, 'getProductById').mockResolvedValue(mockProducts);

            const itemDetails = mount(
                <MemoryRouter>
                    <ItemDetails />
                </MemoryRouter>
            );
            
            expect(itemDetails).toMatchSnapshot();
        });
    });  */

    /* test('Debe renderizar el componente.', () => {
        const itemDetails = mount(
            <MemoryRouter>
                <ItemDetails />
            </MemoryRouter>
        );
        
        console.log(itemDetails.debug())
        expect(itemDetails.length).toEqual(1);
    }) */

    /*  test("Debe renderizar el componente", async () => {
         await act(async () => {
             jest.spyOn(ItemsServices, 'fetchItemById').mockResolvedValue(mockItemsDetail);
 
             const wrapper = mount(<ShowDetail history={mockProperties} />);
             expect(wrapper).toMatchSnapshot();
         });
     }); */
});

describe('ItemDetails sanpshot', () => {
    test('Comprobar UI componente ItemDetails', () => {
        const itemDetails = create(
            <MemoryRouter>
                <ItemDetails />
            </MemoryRouter>
        );
        expect(itemDetails.toJSON()).toMatchSnapshot();
    })
});
