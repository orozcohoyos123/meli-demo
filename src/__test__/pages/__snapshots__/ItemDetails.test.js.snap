// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetails sanpshot Comprobar UI componente ItemDetails 1`] = `
<main
  className="view-wrapper"
>
  <div
    className="view-detail row"
  >
    <div
      className="categories col-11"
    >
      <div
        className="categories__container"
      >
        <a
          href="/items?search="
          onClick={[Function]}
        >
          <span
            className="categories__item bold"
          >
            
          </span>
        </a>
      </div>
    </div>
    <section
      className="product-wrapper col-11"
    >
      <div
        className="row"
      >
        <div
          className="col-12 col-md-8"
        >
          <div>
            <img
              alt=""
              className="product__image"
            />
          </div>
        </div>
        <div
          className="product-buy-info col-12 col-md-4"
        >
          <div
            className="product-sell-quantity"
          >
            <span />
            <span>
                -   
                
            </span>
            <span>
              vendido
              
            </span>
          </div>
          <div
            className="product-name"
          >
            <span />
          </div>
          <div>
            <button
              className="ripple"
            >
              Comprar
            </button>
          </div>
        </div>
      </div>
      <div
        className="product-info-wrapper row"
      >
        <div
          className="product-info-title col-12"
        >
          <span>
            Descripción del producto
          </span>
        </div>
        <div
          className="product-info-detail col-12 col-md-8"
        >
          <span />
        </div>
      </div>
    </section>
  </div>
</main>
`;
